### DISCUSSION TITLE
New tutorial

### DISCUSSION THREAD

**Discussion by awigger** [cross-seed/cross-seed#465] [Ideas]
The tutorial was clearly written by an expert who assumes a much higher floor of knowledge than reality. There are large skips and leaps to next steps that desperately need explanation. Also, a "what to do if you get this error" page or section would also go a long way.

I had to give up installing it on Unraid after an hour of pulling my hair out.

I appreciate the creator(s) for this free and helpful utility, but people who don't write code or know how to read it, would benefit greatly from a much expanded tutorial.

**Comment by mmgoodnow** [comment]
I don't disagree with you, but documentation is really boring to write and this is a passion project for me, so I'm always going to prefer to do what I find fun.

**Comment by RecessedScrew** [comment]
I support the idea of a more newb friendly guide. I've been going in circles for a bit with it. If I figure out how to do it, I'll write something up.

**Comment by RecessedScrew** [comment]
I was able to get it sorted out with the help of some friendly online folks. 

Here are the steps that worked for me to get .torrent file matching working. They are probably redundant or flawed in some way, but they got me there in the end. I may come back and improve the formatting and readability but for now I just wanted to get this out there in hopes it'll help people like me who were stuck. Update: The formatting has been improved and some clarifications made.



The steps listed below were taken on a Windows 10 PC with a local (non-Docker) instance of Prowlarr and a local (non-Docker) instance of qBittorrent:

Install Docker: https://docs.docker.com/desktop/install/windows-install/


### Install Cross-Seed

Open Command Prompt (Win+S > type "CMD" > press enter) or PowerShell, and run the command: docker run crossseed/cross-seed --version
You'll see multiple downloads in process.
Once done, you'll see:

```
Status: Downloaded newer image for crossseed/cross-seed:latest
Configuration file created at /config/config.js
5.4.0
```

In Docker > Containers, a new container is listed with a randomly generated Name (mine was "calm_roberts") and "crossseed/cross-seed:latest" listed under the Image column.


### Create the local host files that Cross-Seed will use

In Windows Explorer, create the folders and docker-compose.yml file:
- Create folder to hold everything: C:\Users\[windows-login-username]\Documents\Docker\cross-seed
- Create config folder: C:\Users\[windows-login-username]\Documents\Docker\cross-seed\config
- Create output folder: C:\Users\[windows-login-username]\Documents\Docker\cross-seed\cross-seeds
- Create yml file: C:\Users\[windows-login-username]\Documents\Docker\cross-seed\docker-compose.yml

To create the folders, right click within the "cross-seed" folder > New > Folder. Name as specified above ("config" and "cross-seeds").

To create the yml file, right click within the "cross-seed" folder > New > Text Document. Name this file "docker-compose.yml" (note: Windows Explorer default settings hide file extensions. To view and edit the extension, in Windows Explorer, go to View > Show/Hide > add a check in the box next to "file name extensions")


### Added content to the yml file

Right-click on docker-compose.yml > Choose "Edit with Notepad++" (if that program is installed, otherwise, Notepad or it may just say "Edit" and you have to choose a compatible program from the provided list).

In the text editor, add the following content (or whatever the latest version is posted in [the tutorial](https://www.cross-seed.org/docs/basics/getting-started) + your modifications to the volumes section):

```
version: "2.1"
services:
    cross-seed:
        image: crossseed/cross-seed
        container_name: cross-seed
        user: 1000:1000 # optional but recommended
        volumes:
            - C:\Users\[windows-login-username]\Documents\Docker\cross-seed\config:/config
            - C:\Users\[windows-login-username]\AppData\Local\qBittorrent\BT_backup:/torrents:ro # note that this volume can and should be mounted read-only
            - C:\Users\[windows-login-username]\Documents\Docker\cross-seed\cross-seeds:/cross-seeds
        command: search
```


### Prepare the torznab URLs

Add the trackers using information from Prowlarr (this assumes your Prowlarr is setup with indexers and running -- Also, see * note at the end of these instructions regarding the indexer URLs):

Log in to Prowlarr
Select Indexers in the left side bar
Choose a tracker and right-click on the RSS icon found on the far right (3rd icon from the right) > Copy Link
Paste into Notepad++ or similar, and you'll see something like this:

	http://localhost:9696/3/api?apikey=sli389asi12908fjh9828hf98hs8h38h&extended=1&t=search

Test the address in a browser. If you see a bunch of text (XML content), then the link is working.


Add quotation marks at each end of the URL:

	"http://localhost:9696/3/api?apikey=sli389asi12908fjh9828hf98hs8h38h&extended=1&t=search"

Repeat those steps for a second indexer, for example (the Prowlarr API key will be the same for each indexer, only the Indexer number changes):

	"http://localhost:9696/4/api?apikey=sli389asi12908fjh9828hf98hs8h38h&extended=1&t=search"

Combine the two, separated by a comma (we'll use these combined URLs in the next step):

	"http://localhost:9696/3/api?apikey=sli389asi12908fjh9828hf98hs8h38h&extended=1&t=search","http://localhost:9696/4/api?apikey=io289jsf98h892398hg98h928h98h238&extended=1&t=search"



### Generate the config.js file

Open Command Prompt (Win+S > type "CMD" > press enter), navigate to the previously created folder containing the yml file:

	cd C:\Users\[windows-login-username]\Documents\Docker\cross-seed\

Start the container:

	docker-compose up

The necessary files will be generated, but the Container will exit because there are no indexers configured.


**Add indexers (torznab URLs) to the Cross-Seed config.js file**

In Windows Explorer, navigate to the Config folder path you previously defined in the yml file. In my case:

	C:\Users\[windows-login-username]\Documents\Docker\cross-seed\config

In the "config" folder, right-click on "config.js" and choose edit (I used Notepad++, which conveniently numbers the rows).

On row 17 (which reads "torznab: [],") between the [], insert the previously combined URL's. Row 17 will look like this after inserting the URLs:

	
	torznab: ["http://localhost:9696/3/api?apikey=sli389asi12908fjh9828hf98hs8h38h&extended=1&t=search","http://localhost:9696/4/api?apikey=io289jsf98h892398hg98h928h98h238&extended=1&t=search"],

Save the document.


### Launch Cross-Seed

Open Command Prompt (Win+S > type "CMD" > press enter), navigate to the previously created folder containing the yml file:

	cd C:\Users\[windows-login-username]\Documents\Docker\cross-seed\

Start the container:

	docker-compose up


If everything worked, you'll see:


[+] Running 1/0
 ✔ Container cross-seed  Created
Attaching to cross-seed
cross-seed  | Configuration file already exists.
cross-seed  | info: Validating your configuration...
cross-seed  | info: Your configuration is valid!
cross-seed  | info: [search] Found 147 torrents, 142 suitable to search for matches


Otherwise, if you see an error, check in the verbose logs, which are found in the previously established folders. In my case, here:

	C:\Users\[windows-login-username]\Documents\Docker\cross-seed\config\logs\verbose.2023-08-30.log






### * A note regarding the torznab URL: 

I'm legitimately confused by my experience with the URLs acquired for the torznab line in the config.js file. There are two points of confusion:

The first point of confusion relates to the format of the URL. Online I see instructions saying to remove the URL content after the API_KEY. Yet when I tested the URL manually in a browser without the "&extended=1&t=search" addition beyond the API_KEY, the page failed to load.

The second point of confusion deals with the IP address. Launching Cross-Seed using "localhost" or "127.0.0.1" in the URL resulted in an error despite both versions of the URL successfully loading torrent details when that URL was manually tested in a browser. 

I ran "ipconfig" in Command Prompt, found the IPv4 Address listed under "Ethernet adapter Ethernet" which was 10.0.0.XXX (where XXX was the three digit final segment of the IP address). Cross-Seed was able to successfully validate and launch using the full URL with this IP address:

`http://10.0.0.XXX:9696/3/api?apikey=sli389asi12908fjh9828hf98hs8h38h&extended=1&t=search`

This all may be due to my running VPN with network lock enabled.

**Comment by RecessedScrew** [comment]
Part II

Next I wanted to get Data Based Matching working. This isn't too difficult with the official tutorial:

https://www.cross-seed.org/docs/tutorials/data-based-matching

However, something on my setup is throwing things off. Following the above tutorial, using Windows paths in the config.js, resulted in "Error: ENOENT: no such file or directory, scandir 'E:\audio'". 

What did work was to setup docker-compose.yml volumes with mappings to my Windows paths:

            - E:\audio:/data/e/audio
            - F:\data\videos:/data/f/data/videos
            - I:\data\videos:/data/i/data/videos

The structure of the path after the ":" is just my chosen pattern. That part can be whatever you'd like (i.e. instead of /data/e/audio it could be /audio or /random/words/and/382i993898numbers)

Then in the config.js, I setup 'dataDirs' like so:

	dataDirs: ["/data/e/audio",
			"/data/f/data/videos",
			"/data/i/data/videos"],

I did the same for 'linkDir', in docker-compose.yml:

            - F:\(data-auto-downloads):/data/auto-downloads

And in config.js:

	linkDir: "/data/auto-downloads",

If I understood the tutorial correctly, avoid nesting the 'linkDir' path within one of the 'dataDirs' paths.

**Reply by akesher123** [reply]
You should be able to escape those paths with a backslash, so in the config have it be "E:\\\audio"

**Comment by zakkarry** [comment]
I took on the challenge, anyone who wants to comment on the PR is welcome.

https://github.com/cross-seed/cross-seed.org/pull/12
